src/drivers/src/EncoderOdometryNode.cpp

```diff
@@ -0,0 +1,290 @@
+    double k_12 = w;
+    double k_20 = v*cos(prev_yaw + dt/2*k_12);
+    double k_21 = v*sin(prev_yaw + dt/2*k_12);
+    double k_22 = w;
+    double k_30 = v*cos(prev_yaw + dt*k_22);
+    double k_31 = v*sin(prev_yaw + dt*k_22);
+    double k_32 = w;
+
+    double x = prev_x + (dt/6) * (k_00 + 2*(k_10 + k_20) + k_30);
+    double y = prev_y + (dt/6) * (k_01 + 2*(k_11 + k_21) + k_31);
+    double yaw = prev_yaw + (dt/6) * (k_02 + 2*(k_12 + k_22) + k_32);
+
+    // Update our estimate
+    last_estimate.pose.pose.position.x = x;
+    last_estimate.pose.pose.position.y = y;
+    tf::quaternionTFToMsg(tf::createQuaternionFromYaw(yaw), last_estimate.pose.pose.orientation);
+    last_estimate.pose.covariance = calculatePoseCovariance(left_encoder_ticks, right_encoder_ticks, prev_yaw);
+    last_estimate.twist.twist.linear.x = x_vel;
+    last_estimate.twist.twist.linear.y = y_vel;
+    last_estimate.twist.twist.angular.z = w;
+    last_estimate.header.stamp = ros::Time::now();
+    last_estimate.header.frame_id = odom_frame_id;
+
+    // Save the current number of ticks as the "previous" number for the
+    // next time we estimate
+    left_encoder_num_ticks_prev = left_encoder_num_ticks_curr;
+    right_encoder_num_ticks_prev = right_encoder_num_ticks_curr;
+
+    // Publish our estimate
+    odom_estimate_publisher.publish(last_estimate);
+}
+
+nav_msgs::Odometry EncoderOdometryNode::generateAllZeroOdometryMessage() {
+    nav_msgs::Odometry odom_msg;
+    odom_msg.pose.pose.position.x = 0;
+    odom_msg.pose.pose.position.y = 0;
+    odom_msg.pose.pose.orientation = tf::createQuaternionMsgFromYaw(0);
+    odom_msg.twist.twist.linear.x = 0;
+    odom_msg.twist.twist.linear.y = 0;
+    odom_msg.twist.twist.linear.z = 0;
+    odom_msg.twist.twist.angular.x = 0;
+    odom_msg.twist.twist.angular.y = 0;
+    odom_msg.twist.twist.angular.z = 0;
+
+    return odom_msg;
+}
+
+geometry_msgs::PoseWithCovariance::_covariance_type EncoderOdometryNode::calculatePoseCovariance(
+        int left_wheel_ticks,
+        int right_wheel_ticks,
+        double prev_yaw) {
+
+    // All math taken from: http://digital.csic.es/bitstream/10261/30337/1/Onto%20computing.pdf
+    // namely pages 1342 and 1343
+
+    // Compute displacement of the left and right wheels (relative to last estimate)
+    double wheel_circumference = 2 * M_PI * wheel_radius;
+    double d_l = left_wheel_ticks * wheel_circumference / ticks_per_rotation;
+    double d_r = right_wheel_ticks * wheel_circumference / ticks_per_rotation;
+
+    // TODO: Detailed doc comments here
+    // let p_t be the change from the previous state (x,y,theta): p_t = [x_t; y_t; theta_t]
+
+    // Cov[p_t] = E[p_t * transpose(p_t)] - E[p_t] * E[transpose(p_t)]
+
+    // Compute E[p_t * transpose(p_t)] - E[p_t]
+    double k_1 = std::pow(d_r, 2) + std::pow(d_l, 2);
+    double k_2 = std::pow(d_r, 2) - std::pow(d_l, 2);
+    double k_3 = d_r * d_l;
+    double k_4 = left_encoder_variance + right_encoder_variance;
```

Comment:
Reality is that all we needed to do was to provide `k_4` and `k_5`

Reply:
Sorry, I'm not sure I totally understand? How would that propagate through the rest of the calculations?

Reply:
Do you mean we would give them to the kalman filter directly?